src/main/java/org/owasp/html/HtmlLexer.java

```diff
@@ -665,6 +679,8 @@ && canonicalElementName(start + 2, end)
                   if ('>' == ch) {
                     state = State.DONE;
                     type = HtmlTokenType.COMMENT;
+                  } else if ('!' == ch) {  // --!> is also valid closing sequence
+                    state = State.COMMENT_DASH_DASH;
```

Comment:
This would seem to suggest that`<!-- --!->` is a whole comment tag.
iiuc, after `<!-- --!-` we should be in the comment_dash state.

Perhaps we need `!` to transition to `COMMENT_DASH_DASH_BANG` here which transitions as does `case COMMENT` above.

Reply:
Does this look right?

```mermaid
flowchart TD
  BANG -- "-" --> BANG_DASH;
  BANG_DASH -- "-" --> COMMENT_DASH_AFTER_BANG;
  BANG_DASH -- "else" --> DIRECTIVE;
  COMMENT_DASH_AFTER_BANG -- "-" --> COMMENT_DASH_AFTER_BANG;
  COMMENT_DASH_AFTER_BANG -- ">" --> DONE;
  COMMENT_DASH_AFTER_BANG -- "else" --> COMMENT;
  COMMENT -- "-" --> COMMENT_DASH;
  COMMENT -- "else" --> COMMENT;
  COMMENT_DASH -- "-" --> COMMENT_DASH_DASH;
  COMMENT_DASH -- "else" --> COMMENT;
  COMMENT_DASH_DASH -- ">" --> DONE;
  COMMENT_DASH_DASH -- "-" --> COMMENT_DASH_DASH;
  COMMENT_DASH_DASH -- "!" --> COMMENT_DASH_DASH_BANG;
  COMMENT_DASH_DASH_BANG -- ">" --> DONE;
  COMMENT_DASH_DASH_BANG -- "-" --> COMMENT_DASH;
  COMMENT_DASH_DASH_BANG -- "else" --> COMMENT;
```

Reply:
Yes, this seems correct. The missing part is the `COMMENT_DASH_AFTER_BANG` case and the `<!>` scenario too. But the new `COMMENT_DASH_DASH_BANG` suggestion seems correct to me.

Reply:
Thanks @mikesamuel , this makes sense. I implemented your proposed changes and added new tests. Everything passes the test. Can you please validate / approve the MR? Thanks!

Reply:
@jfbyers Changes look good to me, let see what mike's thoughts are :)

Reply:
@subbudvk   @mikesamuel  are you ok moving forward with this PR ? What are the next steps? Thank you.